Support for RoAD: 2D Rotary Adaptation

ppetrushkov opened the pull request:
Discussed in issue https://github.com/huggingface/peft/issues/2672

[RoAD](https://arxiv.org/pdf/2409.00119) is a Parameter-efficient Fine-Tuning technique that is especially well-suited for efficient inference with mixed adapters in a batch, while still providing high output quality with very small parameter count.

RoAD learns 2D rotation matrices that are applied to the layer output which can be written using only element-wise multiplication (rather than matrix multiplication), enabling very fast inference with adapters in unmerged state. It is somewhat related to Orthogonal Finetuning (OFT) method.

Paper: https://arxiv.org/pdf/2409.00119

I also have a fork of VLLM where I implemented and tested the efficiency of this method which shows significantly better performance than LoRA https://github.com/ppetrushkov/vllm/tree/v0.9.1-road

BenjaminBossan commented on src/peft/tuners/road/layer.py:
```diff
@@ -0,0 +1,398 @@
+            size = self.out_features * 2
+        else:
+            raise ValueError(f"Unsupported variant {variant} for RoadLayer. Supported variants are 1, 2, and 4.")
+        self.road_theta[adapter_name] = nn.Parameter(torch.rand(size))
+        self.road_alpha[adapter_name] = nn.Parameter(torch.rand(size))
```
WDYT about initializing this with `torch.empty` and then in `reset_road_parameters`, in case that `init_weights=False`, fill them with `rand`? That way, if `init_weights=True` (which is most common), we don't double-initialize the values, making the initialization more efficient.

BenjaminBossan replied:
When passing `init_weights=False`, the values should be randomly initialized in `reset_road_parameters` (which can be renamed to `reset_parameters`) to avoid this.

BenjaminBossan reviewed (changes requested):
Thank you a lot for this PR to add RoAd to PEFT. I have to say, this is one of the most mature PRs to add a new PEFT method so far, really great work on understanding the PEFT code base and following the conventions.

I did a first review pass and didn't find any big flaws. There are a few smaller issues that I found, but I think they should be easily fixed. Please check my comments.

Something I'd be interested in is if you have checked the efficiency of mixed batch inference. If not, I think this it would be great to have a small script to check this. What I would imagine is to load a base model, add multiple RoAd/LoRA adapters (need not be trained), perform single adapter/batch inference and multi adapter/batch inference, then compare the relative overhead. If you don't have such a script, I'd be happy to help working on this.

Once you finish all your changes, don't forget to run `make style` before pushing and then ping me.

ppetrushkov commented on src/peft/tuners/road/config.py:
```diff
@@ -0,0 +1,123 @@
+            however it has some effect on inference speed when used in e.g. VLLM.
+            For best speed group size of at least 64 is recommended.
+            Note that model hidden size (or hidden size per partition when used with tensor parallelism)
+            must be divisible by group_size, so for very small models you might need to reduce this parameter.
```
One issue with large group size that I found appears when using tensor parallelism. Let's say with TP=2, in column parallel case the hidden layer output is split in two halves and each GPU processes its own half. Since computing RoAd requires inputs from the entire group, if we set group_size=hidden_size, it means that both halves need to be available on each GPU, requiring extra synchronisation and code changes. As long as each group fits in a tensor parallel replica, no extra synchronisation is needed.

Smaller group size (like 2), has no such disadvantage in theory, however at least with my current inference implementation it works much faster if the group size is at least 32 or 64.

BenjaminBossan replied:
Okay, got it. I don't think we need to go into quite as much detail in the docs. Let's just mention that 64 is the default:

```diff
For best speed group size of at least 64 is recommended.
For best speed group size of at least 32 or 64 (the default) is recommended.

ppetrushkov commented on src/peft/tuners/road/layer.py:
```diff
@@ -0,0 +1,398 @@
+    """
+    Road layer.
+
+    Generally the idea of RoAD is to split the input vector into many 2D vectors and rotate each 2D vector with its own
```
In this case I'm talking about the theoretical idea of the paper to take a pair of elements from the hidden output, and rotate them with a 2D matrix, so I think saying 2D vector is appropriate. In the actual implementation it is indeed computed as a sum of 2 higher dimensional vectors instead, but in my view this representation is just a mathematical trick.

ppetrushkov commented on src/peft/tuners/road/layer.py:
```diff
@@ -0,0 +1,398 @@
+    x_grouped = x.reshape(-1, 2, group_size // 2)
+    x1 = x_grouped[:, 0, :]
+    x2 = x_grouped[:, 1, :]
+    rotate_half_x = torch.stack((-x2, x1), dim=1).reshape(x.shape)
```
I added a reference to the equation from the paper, basically it comes from the minus sign in the rotation matrix in the top right element.

ppetrushkov commented:
@BenjaminBossan thank you for the thorough review! I addressed some of the comments, will look at the remaining things and get back to you.

One comment regarding benchmarking mixed adapters - my focus was mainly on the VLLM implementation, as this is how we want to use it. To do this efficiently one needs something like grouped gemm to multiply hidden outputs with different adapter weights based on the adapter mapping. I implemented this in VLLM with a triton kernel, somewhat like what punica is doing, but much simpler. Doing this in plain torch would be much slower, or at least I don't know of a good way to do that. So benchmarking the torch version of mixed road inference in peft doesn't make much sense to me. Although, RoAd itself is still much faster than LoRA, whether mixed or not mixed.

FYI attaching one plot I created with my fork of VLLM comparing mixed inference with LoRA vs RoAd vs just base model. 
<img width="1688" height="1361" alt="lora_vs_road_tpot_512_128" src="https://github.com/user-attachments/assets/46e5b1f7-32b9-4936-af5b-a41867d794da" />

BenjaminBossan commented on src/peft/tuners/road/layer.py:
```diff
@@ -0,0 +1,398 @@
+                    self.road_alpha[active_adapter].data,
+                )
+                # Since our matrix are not necessarily orthogonal we need inverse instead of transpose
+                inv_road_R = torch.linalg.inv(road_R.to(torch.float32)).to(orig_dtype)
```
I tested it locally with `meta-llama/Llama-3.1-8B` and unmerging was 4x slower than merging, but still fast overall (0.67 sec). Regarding invertibility, do you think there can be an issue in practice?

ppetrushkov replied:
I thought about invertibility, I think in practice we can basically always expect it to work, since we initialize it as a block-diagonal matrix with each block being a 2D rotation matrix (but the rows are permuted when group_size>2) and from what I've seen after training the values don't change very significantly.

BenjaminBossan replied:
Sounds good, then let's just leave this as is.

BenjaminBossan reviewed (commented):
Regarding benchmarks: It makes sense that users who would be searching for the best performance would look into vLLM etc. Still, I think it's good to check if the advantages also manifest in PEFT.

I created a small toy benchmark:

<details>

<summary>Show code</summary>

```python
import time

import torch
from transformers import AutoModelForCausalLM
from peft import get_peft_model, RoadConfig, LoraConfig

num_adapters = 10
num_iters = 50
model_id = "meta-llama/Llama-3.2-3B"

config0 = LoraConfig(init_lora_weights=False, r=4)  # matching params of RoadConfig
config1 = RoadConfig(init_weights=False)
inputs = [torch.randint(0, 100, (num_adapters, 50)).to(0) for _ in range(num_iters)]

for config in [config0, config1]:
    model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16).to(0)
    model = get_peft_model(model, config, adapter_name="adapter0")
    for i in range(1, num_adapters):
        model.add_adapter(f"adapter{i}", config)
    model.eval()

    adapter_names = [f"adapter{i}" for i in range(num_adapters)]

    # warmup
    model(inputs[0])

    # Run the model w/o mixed adapter batch
    durations = {"no mixed batch": [], "   mixed batch": []}

    for i in range(num_iters):
        x = inputs[i]
        tic = time.perf_counter()
        model(x)
        toc = time.perf_counter()
        durations["no mixed batch"].append(toc - tic)

    # Run the model with mixed adapter batch
    for i in range(num_iters):
        x = inputs[i]
        tic = time.perf_counter()
        model(x, adapter_names=adapter_names)
        toc = time.perf_counter()
        durations["   mixed batch"].append(toc - tic)

    print(f"{config.peft_type.value} summary (over 10 runs)")
    for name, dur in durations.items():
        print(f"{name}: avg {sum(dur)/num_iters:.4f} sec, min {min(dur):.4f} sec, max {max(dur):.4f} sec, total: {sum(dur):.4f} sec")

    del model
    torch.cuda.empty_cache()
```

</details>

The results look quite promising so far:

```
LORA summary (over 10 runs)
no mixed batch: avg 0.0262 sec, min 0.0159 sec, max 0.0272 sec, total: 1.3117 sec
   mixed batch: avg 0.0878 sec, min 0.0848 sec, max 0.1182 sec, total: 4.3920 sec
ROAD summary (over 10 runs)
no mixed batch: avg 0.0263 sec, min 0.0179 sec, max 0.0269 sec, total: 1.3160 sec
   mixed batch: avg 0.0266 sec, min 0.0263 sec, max 0.0269 sec, total: 1.3297 sec
```

ppetrushkov commented:
@BenjaminBossan I think I addressed all your comments so far, please have another look. Thanks!

BenjaminBossan commented on examples/road_finetuning/road_finetuning.py:
Not a huge deal, but I got an OOM with 24 GB VRAM using the script like this:

`python road_finetuning.py --base_model "meta-llama/Llama-3.2-3B" --data_path timdettmers/openassistant-guanaco --variant road_1`

The same script runs when using LoRA (rank 8) instead of RoAD. I had to restrict the target layers to `--road_target_modules "q_proj,v_proj"` to avoid OOM.

BenjaminBossan replied:
Makes sense, thanks for explaining.

BenjaminBossan reviewed (changes requested):
Thanks for addressing my comments, there is not much missing. There is an error in the included example, please check my comments. The of my comments are just smaller changes.

Before we can merge, since we have bnb support, let's also add some GPU tests for this. You can copy the RandLoRA examples and adjust them for RoAD:

https://github.com/huggingface/peft/blob/e0b2ca79771066162d6efec86764b047a323b98a/tests/test_gpu_examples.py#L1492-L1718

ppetrushkov commented:
@BenjaminBossan done.

Regarding MetaMath, I tried a bit, but didn't improve much. Including all linear layers helped most. I think some tasks that require larger changes and more parameters are harder for RoAd.

BenjaminBossan reviewed (changes requested):
Thanks for all the changes, this PR looks pretty much ready for me. Please run `make style` to ensure that the linter is happy.

ppetrushkov commented:
@BenjaminBossan added it to the toc, let me know if anything else is missing.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2678). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
